helpers/browser/driver.js

```diff
@@ -187,37 +180,57 @@ class ChromeProtocol {
   }
 
   beginTrace() {
-    this._traceEvents = [];
     const tracingOpts = {
       categories: this._traceCategories.join(','),
+      transferMode: 'ReturnAsStream',
       options: 'sampling-frequency=10000'  // 1000 is default and too slow.
     };
 
-    this.on('Tracing.dataCollected', data => {
-      this._traceEvents.push(...data.value);
-    });
-
     return this.connect()
       .then(_ => this.sendCommand('Page.enable'))
       .then(_ => this.sendCommand('Tracing.start', tracingOpts));
   }
 
   endTrace() {
-    return this.connect().then(_ => {
-      return new Promise((resolve, reject) => {
-        // Limit trace retrieval execution time.
-        const traceTimeoutId = setTimeout(_ => {
-          reject(new Error('Trace retrieval timed out'));
-        }, TRACE_RETRIEVAL_TIMEOUT);
-
-        // When all Tracing.dataCollected events have finished, this event fires.
-        this.on('Tracing.tracingComplete', _ => {
-          clearTimeout(traceTimeoutId);
-          resolve(this._traceEvents);
-        });
-
-        this.sendCommand('Tracing.end');
+    return new Promise((resolve, reject) => {
+      // When the tracing has ended this will fire with a stream handle.
+      this.on('Tracing.tracingComplete', streamHandle => {
+        this._readTraceFromStream(streamHandle)
+            .then(traceContents => resolve(traceContents));
       });
+
+      // Issue the command to stop tracing.
+      this.connect().then(_ => this.sendCommand('Tracing.end'));
```

Comment:
at some point we're going to have to clean up this `connect()` business, since `this.on` will have already thrown before this if `this.connect()` doesn't immediately resolve. Right now, technically no `this.connect` is necessary, since the runner has already called `driver.connect` before making any subsequent calls, but we might not want to require clients of the driver to have to know that. We just have to decide where we want the guard to live.